plugins/intel-usb4/fu-intel-usb4-device.c

```diff
@@ -429,196 +420,141 @@ fu_dell_dock_usb4_activate(FuDevice *device, FuProgress *progress, GError **erro
-	fu_device_set_version(device, fw_version);
+	/* success */
 	return TRUE;
 }
 
 static gboolean
-fu_dell_dock_usb4_setup(FuDevice *device, GError **error)
+fu_intel_usb4_device_setup(FuDevice *device, GError **error)
 {
-	guchar buf[NVM_READ_LENGTH] = {0x0};
+	FuIntelUsb4Device *self = FU_INTEL_USB4_DEVICE(device);
+	guint8 buf[NVM_READ_LENGTH] = {0x0};
 	g_autofree gchar *name = NULL;
-	g_autofree gchar *nvm_product_id = NULL;
-	g_autofree gchar *nvm_vendor_id = NULL;
-	g_autofree gchar *nvm_version = NULL;
+	g_autoptr(FuFirmware) fw = fu_intel_thunderbolt_nvm_new();
+	g_autoptr(GBytes) blob = NULL;
 
-	if (!fu_dell_dock_usb4_hub_nvm_read(device, buf, NVM_READ_LENGTH, 0, error)) {
-		g_prefix_error(error, "NVM READ error: ");
+	/* read from device and parse firmware */
+	if (!fu_intel_usb4_device_nvm_read(device, buf, sizeof(buf), 0, error)) {
+		g_prefix_error(error, "NVM read error: ");
 		return FALSE;
 	}
-	nvm_version =
-	    g_strdup_printf("%02x.%02x", buf[NVM_VER_OFFSET_MAJOR], buf[NVM_VER_OFFSET_MINOR]),
-	nvm_vendor_id =
-	    g_strdup_printf("%02x%02x", buf[NVM_VID_OFFSET_MAJOR], buf[NVM_VID_OFFSET_MINOR]);
-	nvm_product_id =
-	    g_strdup_printf("%02x%02x", buf[NVM_PID_OFFSET_MAJOR], buf[NVM_PID_OFFSET_MINOR]);
-
-	/* only add known supported thunderbolt devices */
-	name = g_strdup_printf("TBT-%s%s", nvm_vendor_id, nvm_product_id);
-	if (g_strcmp0(name, DELL_DOCK_USB4_INSTANCE_ID) != 0) {
-		g_set_error_literal(error,
-				    FWUPD_ERROR,
-				    FWUPD_ERROR_NOT_SUPPORTED,
-				    "no supported device found");
+	blob = g_bytes_new(buf, sizeof(buf));
+	if (!fu_firmware_parse(fw, blob, FWUPD_INSTALL_FLAG_NONE, error)) {
+		g_prefix_error(error, "NVM parse error: ");
 		return FALSE;
 	}
-	fu_device_add_instance_id(device, name);
-	fu_device_set_version_format(device, FWUPD_VERSION_FORMAT_PAIR);
-	fu_device_set_version(device, nvm_version);
-	return TRUE;
-}
+	self->nvm_vendor_id = fu_intel_thunderbolt_nvm_get_vendor_id(FU_INTEL_THUNDERBOLT_NVM(fw));
+	self->nvm_product_id = fu_intel_thunderbolt_nvm_get_model_id(FU_INTEL_THUNDERBOLT_NVM(fw));
 
-static gboolean
-fu_dell_dock_usb4_probe(FuDevice *device, GError **error)
-{
-	FuDellDockUsb4 *self = FU_DELL_DOCK_USB4(device);
-
-	self->intf_nr = GR_USB_INTERFACE_NUMBER;
-	self->blocksz = GR_USB_BLOCK_SIZE;
-	fu_device_set_logical_id(FU_DEVICE(device), "usb4");
+	name = g_strdup_printf("TBT-%04x%04x", self->nvm_vendor_id, self->nvm_product_id);
+	fu_device_add_instance_id(device, name);
+	fu_device_set_version(device, fu_firmware_get_version(fw));
 	return TRUE;
 }
 
 static void
-fu_dell_dock_usb4_finalize(GObject *object)
+fu_intel_usb4_device_to_string(FuDevice *device, guint idt, GString *str)
 {
-	G_OBJECT_CLASS(fu_dell_dock_usb4_parent_class)->finalize(object);
+	FuIntelUsb4Device *self = FU_INTEL_USB4_DEVICE(device);
+	fu_string_append_kx(str, idt, "NvmVendorId", self->nvm_vendor_id);
+	fu_string_append_kx(str, idt, "NvmProductId", self->nvm_product_id);
```

Comment:
I don't think there is a lot of value to this vfunc.  The product and vendor are put into the instance ID already.

Reply:
It was literally added just for debugging, rather than mentally parsing the `TBT-%04x%04x` ID.